Sources/CRC/CRC32.swift

```diff
@@ -40,16 +40,90 @@ struct CRC32:Hashable, Sendable
         return checksum
     }
 
+    /// Returns a new checksum by hashing the provided message into the current checksum.
+    ///
+    /// This manually specialized implementation is much faster in debug mode than the
+    /// generic implementation, but exactly the same in release mode.
+    @inlinable public
+    func updated(with message:borrowing [UInt8]) -> Self
+    {
+        var checksum:Self = self
+        checksum.update(with: message)
+        return checksum
+    }
+
     /// Updates the checksum by hashing the provided message into the existing checksum.
     @inlinable public mutating
     func update(with message:borrowing some Sequence<UInt8>)
     {
         self.checksum = ~message.reduce(~self.checksum)
         {
             (state:UInt32, byte:UInt8) in
-            Self.table[Int.init(UInt8.init(truncatingIfNeeded: state) ^ byte)] ^ state >> 8
+            let indexByte:UInt8 = UInt8.init(truncatingIfNeeded: state) ^ byte
+            let index:Int
+            #if DEBUG
+                // in debug mode these hacky integer conversions make this function
+                // around 35% faster
+                if MemoryLayout<Int>.stride == 8 {
+                    let tuple:(UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8) =
+                    (
+                        indexByte, 0, 0, 0, 0, 0, 0, 0
+                    )
+                    index = unsafeBitCast(tuple, to: Int.self)
+                } else {
+                    let tuple:(UInt8, UInt8, UInt8, UInt8) =
+                    (
+                        indexByte, 0, 0, 0
+                    )
+                    index = unsafeBitCast(tuple, to: Int.self)
+                }
+            #else
+                index = Int.init(indexByte)
+            #endif
+            return Self.table[index] ^ state >> 8
         }
     }
+
```

Comment:
out of curiosity, can we get away with a `[UInt8]` typed overload that calls into the generic `Sequence<UInt8>` overload?

this ought to be enough to trigger specialization if i remember correctly

Reply:
This is specifically a debug mode optimisation, and I don't think the compiler really does any specialisation in debug mode. Optimising for debug mode and optimising for release mode seem to be completely different beasts. As soon as a generic function is called, even if it's in the standard library, there seems to be a bunch of overhead. Here of course the overhead isn't at the call-site of that function, but it accumulates on every step of reduce due to the generic iterator and stuff.

Reply:
okay